skimage/morphology/_util.py

```diff
@@ -51,35 +54,206 @@ def _validate_connectivity(image_dim, connectivity, offset):
+        # np.unique reorders, which we don't want
+        _, indices = np.unique(raveled_offsets, return_index=True)
+        raveled_offsets = raveled_offsets[np.sort(indices)]
+
+    # Remove "offset to center"
+    raveled_offsets = raveled_offsets[1:]
+
+    return raveled_offsets
+
+
+def _resolve_neighborhood(selem, connectivity, ndim):
+    """Validate or create structuring element.
+
+    Depending on the values of `connectivity` and `selem` this function
+    either creates a new structuring element (`selem` is None) using
+    `connectivity` or validates the given structuring element (`selem` is not
+    None).
+
+    Parameters
+    ----------
+    selem : ndarray, optional
+        A structuring element used to determine the neighborhood of each
+        evaluated pixel (``True`` denotes a connected pixel). It must be a
+        boolean array and have the same number of dimensions as `image`. If
+        neither `selem` nor `connectivity` are given, all adjacent pixels are
+        considered as part of the neighborhood.
+    connectivity : int, optional
+        A number used to determine the neighborhood of each evaluated pixel.
+        Adjacent pixels whose squared distance from the center is less than or
+        equal to `connectivity` are considered neighbors. Ignored if
+        `selem` is not None.
+    ndim : int
+        Number of dimensions `selem` ought to have.
+
+    Returns
+    -------
+    selem : ndarray
+        Validated or new structuring element specifying the neighborhood.
+
+    Examples
+    --------
+    >>> _resolve_neighborhood(None, 1, 2)
+    array([[False,  True, False],
+           [ True,  True,  True],
+           [False,  True, False]])
+    >>> _resolve_neighborhood(None, None, 3).shape
+    (3, 3, 3)
     """
-    structure = structure.copy()  # Don't modify original input
-    structure[tuple(center)] = 0  # Ignore the center; it's not a neighbor
-    connection_indices = np.transpose(np.nonzero(structure))
-    offsets = (np.ravel_multi_index(connection_indices.T, image_shape,
-                                    order=order) -
-               np.ravel_multi_index(center, image_shape, order=order))
-    squared_distances = np.sum((connection_indices - center) ** 2, axis=1)
-    return offsets[np.argsort(squared_distances)]
+    if selem is None:
+        if connectivity is None:
+            connectivity = ndim
+        selem = ndi.generate_binary_structure(ndim, connectivity)
+    else:
+        # Validate custom structured element
+        selem = np.asarray(selem, dtype=np.bool)
+        # Must specify neighbors for all dimensions
+        if selem.ndim != ndim:
+            raise ValueError(
+                "structuring element and image must have the same number of "
+                "dimensions"
+            )
+        # Must only specify direct neighbors
+        if any(s != 3 for s in selem.shape):
+            raise ValueError("dimension size in structuring element is not 3")
```

Comment:
I recently wanted to specify a structuring element as a flat disk, but was forced by scipy.ndimage to pad it with zeros on the flat axis, which I found annoying. I wonder if, for the use cases envisioned by this function, we can relax this requirement...

Reply:
You are suggesting that this function would automatically replace `selem=np.array([[1, 1, 1]])` with `selem=np.array([[0, 0, 0], [1, 1, 1], [0, 0, 0]])`? I'm not convinced that the convenience is worth it. 

- The behavior for different values to `selem` and `connectivity` already requires extensive checks and documentation without adding the suggested behavior.
- The suggested behavior seems a little arbitrary. Even though it makes sense for your given example, another user may expect `[[1, 1, 1], [1, 1, 1], [1, 1, 1]]`.

I'd rather be explicit to avoid as much confusion as possible.

Reply:
I disagree. The behaviour of such a footprint is perfectly well defined in the context of convolution. This would be exactly analogous, *and*, I’m not asking for it to be expanded into anything, only to be allowed. The offsets function will work perfectly fine with that syntax.

Reply:
Ah, that is worth thinking about. Sorry for misunderstanding. Currently we expect all dimensions in `selem` to be of length 3 and assume that the center is always at `(1,) * ndim` (e.g. in `flood_fill`, `local_maxima`). In `watershed` and `max_tree` `_validate_connectivity` is used which raises if there is an ambiguous center (`selem` has dimension with even length). That might work here as well. But I'd prefer to do that in a separate PR...

Reply:
Ok, I suppose adding more permissive behaviour in the future is totally fine. =)

Reply:
I opened #4567 as a reminder.